proxy/proxy.go

```diff
@@ -871,7 +871,11 @@ func (p *Proxy) makeBackendRequest(ctx *context) (*http.Response, *proxyError) {
 		p.log.Errorf("could not map backend request, caused by: %v", err)
 		return nil, &proxyError{err: err}
 	}
-
+	if ctx.route.BackendType == eskip.LBBackend && ctx.route.LBEndpoints != nil {
+		endpoint := ctx.route.LBEndpoints.Get(req.URL.Host)
```

Comment:
Alternatively chosen endpoint could be returned from mapRequest to avoid creation of a new `LBEndpointCollection` datastructure.

Reply:
I agree this is not elegant. We discussed a bunch of options regarding how to handle the metrics collection for LBs with @aryszka . I tried to keep the code as less intrusive as possible otherwise, the refactoring would take so much.

Reply:
> ... the refactoring would take so much

```diff
diff --git a/proxy/proxy.go b/proxy/proxy.go
index 58f7115..f65fedb 100644
--- a/proxy/proxy.go
+++ b/proxy/proxy.go
@@ -470,22 +470,20 @@ func setRequestURLForDynamicBackend(u *url.URL, stateBag map[string]interface{})
        }
 }
 
-func setRequestURLForLoadBalancedBackend(u *url.URL, rt *routing.Route, lbctx *routing.LBContext) {
-       e := rt.LBAlgorithm.Apply(lbctx)
-       u.Scheme = e.Scheme
-       u.Host = e.Host
-}
-
 // creates an outgoing http request to be forwarded to the route endpoint
 // based on the augmented incoming request
-func mapRequest(r *http.Request, rt *routing.Route, host string, removeHopHeaders bool, stateBag map[string]interface{}) (*http.Request, error) {
+func mapRequest(r *http.Request, rt *routing.Route, host string, removeHopHeaders bool, stateBag map[string]interface{}) (*http.Request, *routing.LBEndpoint, error) {
+       var endpoint *routing.LBEndpoint
        u := r.URL
        switch rt.BackendType {
        case eskip.DynamicBackend:
                setRequestURLFromRequest(u, r)
                setRequestURLForDynamicBackend(u, stateBag)
        case eskip.LBBackend:
-               setRequestURLForLoadBalancedBackend(u, rt, routing.NewLBContext(r, rt))
+               e := rt.LBAlgorithm.Apply(routing.NewLBContext(r, rt))
+               u.Scheme = e.Scheme
+               u.Host = e.Host
+               endpoint = &e
        default:
                u.Scheme = rt.Scheme
                u.Host = rt.Host
@@ -498,7 +496,7 @@ func mapRequest(r *http.Request, rt *routing.Route, host string, removeHopHeader
 
        rr, err := http.NewRequest(r.Method, u.String(), body)
        if err != nil {
-               return nil, err
+               return nil, endpoint, err
        }
 
        rr = rr.WithContext(r.Context())
@@ -527,7 +525,7 @@ func mapRequest(r *http.Request, rt *routing.Route, host string, removeHopHeader
                rr = rr.WithContext(ot.ContextWithSpan(rr.Context(), ctxspan))
        }
 
-       return rr, nil
+       return rr, endpoint, nil
 }
 
 func forwardToProxy(incoming, outgoing *http.Request) {
@@ -866,13 +864,12 @@ func (p *Proxy) makeUpgradeRequest(ctx *context, req *http.Request) error {
 
 func (p *Proxy) makeBackendRequest(ctx *context) (*http.Response, *proxyError) {
        var err error
-       req, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost, p.flags.HopHeadersRemoval(), ctx.StateBag())
+       req, endpoint, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost, p.flags.HopHeadersRemoval(), ctx.StateBag())
        if err != nil {
                p.log.Errorf("could not map backend request, caused by: %v", err)
                return nil, &proxyError{err: err}
        }
-       if ctx.route.BackendType == eskip.LBBackend && ctx.route.LBEndpoints != nil {
-               endpoint := ctx.route.LBEndpoints.Get(req.URL.Host)
+       if endpoint != nil {
                endpoint.Metrics.IncInflightRequest()
                defer endpoint.Metrics.DecInflightRequest()
        }
@@ -1129,7 +1126,7 @@ func (p *Proxy) do(ctx *context) error {
                ctx.setResponse(loopCTX.response, p.flags.PreserveOriginal())
                ctx.proxySpan = loopCTX.proxySpan
        } else if p.flags.Debug() {
-               debugReq, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost, p.flags.HopHeadersRemoval(), ctx.StateBag())
+               debugReq, _, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost, p.flags.HopHeadersRemoval(), ctx.StateBag())
                if err != nil {
                        return &proxyError{err: err}
                }
```

Reply:
The change from above will not work because Algorithm.Apply returns endpoint value so later modification of inflight request counter would modify the copy.

Reply:
Today I am testing the algorithm with some synthetic load tests. The diff you posted looks good to me. I think we can applied once I prove it is valuable and brings value.